### PR TITLE
feat(diagnostics): support multiline diagnostics

### PR DESCRIPTION
**Pull request by 0xLucqs**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)

### CODE REVIEW

File: crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors [hunk 1]
```diff
@@ -1265,8 +1265,8 @@ TerminalSemicolon
 //! > expected_diagnostics
 error: Skipped tokens. Expected: Const/Enum/ExternFunction/ExternType/Function/Impl/InlineMacro/Module/Struct/Trait/TypeAlias/Use or an attribute.
  --> dummy_file.cairo:2:7
-#[bbb]
-      ^
+\   macro
```
Comment (0xLucqs) [inline crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors:1268]: @orizi any idea what's happening there ?

File: crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors [hunk 2]
```diff
@@ -1265,8 +1265,9 @@ TerminalSemicolon
 //! > expected_diagnostics
 error: Skipped tokens. Expected: Const/Enum/ExternFunction/ExternType/Function/Impl/InlineMacro/Module/Struct/Trait/TypeAlias/Use or an attribute.
  --> dummy_file.cairo:2:7
-#[bbb]
-      ^
+\   
+|   macro
```
Comment (0xLucqs) [inline crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors:1269]: @orizi any idea what's happening there ?

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 57 files at r1, 1 of 22 files at r2, 3 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 5 of 57 files reviewed, 3 unresolved discussions (waiting on @0xLucqs and @ilyalesokhin-starkware)

___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 20 at r3](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3B3HCO-k7bIkWzZGuG:-O3B3HCO-k7bIkWzZGuH:b-miat0n) ([raw file](https://github.com/starkware-libs/cairo/blob/c16790ada7d94276c09e29acf24663a16baccd54/crates/cairo-lang-diagnostics/src/location_marks.rs#L20)):*
> ```XML
> 
>     let mut span = location.span;
>     span.end = std::cmp::min(summary.last_offset, span.end);
> ```

doc what you are doing here.



_Code quote:_
```XML
    let mut span = location.span;
    span.end = std::cmp::min(summary.last_offset, span.end);
```
___
*[`crates/cairo-lang-diagnostics/src/location_marks_test.rs` line 35 at r3](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3B3j_3BZku8KrKEteS:-O3B3j_3BZku8KrKEteT:b-axtmtw) ([raw file](https://github.com/starkware-libs/cairo/blob/c16790ada7d94276c09e29acf24663a16baccd54/crates/cairo-lang-diagnostics/src/location_marks_test.rs#L35)):*
> ```XML
>             someotherevenlongerfunctionnamewhichisveryveryveryveryverylong()
>         };
>         Scope is closed."};
> ```

no need to make this actually code based - it actually makes looking at it more complicated.

also test a single char diagnostics at the end of a line.



_Code quote:_
```XML
        match something {
            This is a new scope
        };
        let a = SomeStruct { some_value };
        let b = if true { a } else { b };
        let a_long_variable_name_to_go_to_newline = if someverylongbooleanvalue {
            if someverylongfunctionname() {
                some_value
            } else {
                some_other_value
            }
        } else {
            someotherevenlongerfunctionnamewhichisveryveryveryveryverylong()
        };
        Scope is closed."};
```

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 3 of 57 files at r1, 2 of 26 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 7 of 57 files reviewed, 5 unresolved discussions (waiting on @0xLucqs and @ilyalesokhin-starkware)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3Cdf5f9Qizis38T2Rd:-O3Cdf5f9Qizis38T2Re:bmmere) (no related file):*
@ilyalesokhin-starkware 

___
___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 21 at r4](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CcPh1AdIphbgQBdb-:-O3CcPh1AdIphbgQBdb0:b-aeczxr) ([raw file](https://github.com/starkware-libs/cairo/blob/65d7093ea46e7419d93c718658434b640578b576/crates/cairo-lang-diagnostics/src/location_marks.rs#L21)):*
> ```XML
>     let mut span = location.span;
>     // Ensure the span does not go beyond the end of the file.
>     span.end = std::cmp::min(summary.last_offset, span.end);
> ```

is `last` and `end` equivalent? or do they have an offset of 1?



_Code quote:_
```XML
    span.end = std::cmp::min(summary.last_offset, span.end);
```
___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 117 at r4](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3Cclvl5AxAS-VvNW3w:-O3Cclvl5AxAS-VvNW3x:bxiug8g) ([raw file](https://github.com/starkware-libs/cairo/blob/65d7093ea46e7419d93c718658434b640578b576/crates/cairo-lang-diagnostics/src/location_marks.rs#L117)):*
> ```XML
>     res
> }
> pub fn get_location_marks(
> ```

empty line between functions.

put helpers after actual function (so move it to top)

and probably doc if you are already here.



_Suggestion:_
```XML
}

pub fn get_location_marks(
```
___
*[`crates/cairo-lang-lowering/src/test.rs` line 135 at r5](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CdURz6XcWVkF9vP-r:-O3CdURz6XcWVkF9vP-s:b-6j3go2) ([raw file](https://github.com/starkware-libs/cairo/blob/9eca9b262c7808597719958ffb268fb4f685fbb2/crates/cairo-lang-lowering/src/test.rs#L135)):*
> ```XML
> |       a = a * 3
> |   }; }
> |______^
> ```

should it be something like:
?
i don't really understand the formatting.



_Suggestion:_
```XML
\\   fn test_func() { let mut a = 5; {
||      a = a * 3
||  }; }
|___^__^
```
___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors` line 1269 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6105#gh-1699146891:-O3Cdq8YCdMTZ7GrOzWp:bq5n9xu) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0ecd7e9b8ee55641e7ae5481cf71e650e528c/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors#L1269)):*
<details><summary><i>Previously, 0xLucqs (0xLucqs) wrote…</i></summary><blockquote>

@orizi any idea what's happening there ? (does the diagnostic look correct ?)
</blockquote></details>

it seems like you have a bug with EOL only diags.

**Review by 0xLucqs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 7 of 57 files reviewed, 5 unresolved discussions (waiting on @ilyalesokhin-starkware and @orizi)

___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 21 at r4](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CcPh1AdIphbgQBdb-:-O3CmM1-1Pq-Gq1n0J8C:b-u4wnrh) ([raw file](https://github.com/starkware-libs/cairo/blob/65d7093ea46e7419d93c718658434b640578b576/crates/cairo-lang-diagnostics/src/location_marks.rs#L21)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

is `last` and `end` equivalent? or do they have an offset of 1?
</blockquote></details>

`span.end` is what we get as input and `summary.last_offset` is the last offset of the file, this is to make sure that our span doesn't go beyond the end of the file
___
*[`crates/cairo-lang-lowering/src/test.rs` line 135 at r5](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CdURz6XcWVkF9vP-r:-O3CoHzT5lUpfCwUYU2K:b-axyu59) ([raw file](https://github.com/starkware-libs/cairo/blob/9eca9b262c7808597719958ffb268fb4f685fbb2/crates/cairo-lang-lowering/src/test.rs#L135)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

should it be something like:
?
i don't really understand the formatting.
</blockquote></details>

no this is correct, the span starts at the beginning of the function definition and ends at the end. The weird formatting of the function itself comes from the definition here `let function_code: String = format!("fn test_func() {{ {function_body} {{\n{expr_code}\n}}; }}");`
___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors` line 1269 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6105#gh-1699146891:-O3CoeMe6eECs8FfDEKF:bz6npk3) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0ecd7e9b8ee55641e7ae5481cf71e650e528c/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors#L1269)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

it seems like you have a bug with EOL only diags.
</blockquote></details>

what should be displayed there ?

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 7 of 57 files reviewed, 4 unresolved discussions (waiting on @0xLucqs and @ilyalesokhin-starkware)

___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 21 at r4](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CcPh1AdIphbgQBdb-:-O3CrJRdAYTi_k6NRE-w:b-w2r0kw) ([raw file](https://github.com/starkware-libs/cairo/blob/65d7093ea46e7419d93c718658434b640578b576/crates/cairo-lang-diagnostics/src/location_marks.rs#L21)):*
<details><summary><i>Previously, 0xLucqs (0xLucqs) wrote…</i></summary><blockquote>

`span.end` is what we get as input and `summary.last_offset` is the last offset of the file, this is to make sure that our span doesn't go beyond the end of the file
</blockquote></details>

i think `.end` is exclusive and `last_offset` is inclusive.
please check.
___
*[`crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors` line 1269 at r2](https://reviewable.io/reviews/starkware-libs/cairo/6105#gh-1699146891:-O3CrXfT7LLKx_jf_4RH:b-ssj297) ([raw file](https://github.com/starkware-libs/cairo/blob/92c0ecd7e9b8ee55641e7ae5481cf71e650e528c/crates/cairo-lang-parser/src/parser_test_data/partial_trees_with_trivia/attribute_errors#L1269)):*
<details><summary><i>Previously, 0xLucqs (0xLucqs) wrote…</i></summary><blockquote>

what should be displayed there ?
</blockquote></details>

exactly what was before.

**Review by 0xLucqs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 7 of 57 files reviewed, 4 unresolved discussions (waiting on @ilyalesokhin-starkware and @orizi)

___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 21 at r4](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3CcPh1AdIphbgQBdb-:-O3DEMMe3HltnOX4IkRH:b-1wh3cx) ([raw file](https://github.com/starkware-libs/cairo/blob/65d7093ea46e7419d93c718658434b640578b576/crates/cairo-lang-diagnostics/src/location_marks.rs#L21)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

i think `.end` is exclusive and `last_offset` is inclusive.
please check.
</blockquote></details>

yes it's the case but that's good because that's what is done:
`content[(self.start.0.0 as usize)..(self.end.0.0 as usize)]` so if `.end` is `.last_offset + 1` this will panic

**Review by mkaput — CHANGES_REQUESTED** [review]
> Reviewed 1 of 55 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 5 of 71 files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

___
*[`crates/cairo-lang-diagnostics/src/diagnostics.rs` line 61 at r6](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3DSk2R04dWS3KyW0_f:-O3DSk2R04dWS3KyW0_g:bujo0uo) ([raw file](https://github.com/starkware-libs/cairo/blob/7d8243fd2f1f6790badc9bb3005971889ea049a4/crates/cairo-lang-diagnostics/src/diagnostics.rs#L61)):*
> ```XML
>     pub file_id: FileId,
>     pub span: TextSpan,
>     pub severity: Option<Severity>,
> ```

But severity has nothing to do with a location, isn't it?

Shouldn't this property be passed **along** the location instead of **within** it?

**Review by mkaput — COMMENTED** [review]
> Reviewed 4 of 57 files at r1, 1 of 22 files at r2, 2 of 6 files at r3, 9 of 26 files at r4, 33 of 55 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 49 of 71 files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

**Review by mkaput — COMMENTED** [review]
> Reviewed 1 of 6 files at r3, 4 of 55 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 54 of 71 files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

**Review by mkaput — COMMENTED** [review]
> Reviewed 1 of 55 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 55 of 71 files reviewed, 6 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3DUaZN09_hjD2N2obX:-O3DUaZN09_hjD2N2obY:bccyenf) (no related file):*
At this point, I strongly suggest reusing one of the existing libraries (and their UXes) instead of reinventing stuff in-house. Just look how much code you had to change, and how much complexity got added.

For example, Rust people have their diagnostics crate public, and it looks like it does exactly what you are attempting here: https://github.com/rust-lang/annotate-snippets-rs

___

**Review by mkaput — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: 55 of 71 files reviewed, 7 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

___
*[`crates/cairo-lang-diagnostics/src/location_marks.rs` line 80 at r6](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3DUzBx-StcAO0SXoMs:-O3DUzBx-StcAO0SXoMt:b-atrno1) ([raw file](https://github.com/starkware-libs/cairo/blob/7d8243fd2f1f6790badc9bb3005971889ea049a4/crates/cairo-lang-diagnostics/src/location_marks.rs#L80)):*
> ```XML
>         None => |val: &str| Colorize::normal(val),
>         Some(Severity::Warning) => |val: &str| Colorize::bold(Colorize::yellow(val)),
>         Some(Severity::Error) => |val: &str| Colorize::bold(Colorize::bright_red(val)),
> ```

Unfortunately, you cannot colourize diagnostics this early because you will break Scarb

Scarb **has** to get naked diagnostic text that can be colourized in a separate post-process phase.  This is needed because Scarb is printing either raw or colorized diagnostics depending on runtime context (it has JSON output, supports NO_COLOR envs, handles non-TTY stdout differently).

**Comment by 0xLucqs** [comment]
Removed everything about colors and fixed the macro thing. Agree with @mkaput this crate looks amazing and will remove a huge burden for diagnostics

**Review by mkaput — APPROVED** [review]
:lgtm:

> Reviewed 56 of 56 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105#-:-O3HGRKP4JwwuTrKw3zJ:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

___
*[`crates/cairo-lang-diagnostics/src/diagnostics_test.rs` line 68 at r7](https://reviewable.io/reviews/starkware-libs/cairo/6105#-O3HCK0J1IwbK1cS9Jo1:-O3HCK0J1IwbK1cS9Jo2:b2nhei9) ([raw file](https://github.com/starkware-libs/cairo/blob/4ee55080dfde82b573458dd9f4314b4aba2c5273/crates/cairo-lang-diagnostics/src/diagnostics_test.rs#L68)):*
> ```XML
>             |___^
> 
>         " }
> ```

FYI



_Suggestion:_
```XML
        indoc! {r#"
            error: Simple diagnostic.
             --> dummy_file.sierra:1:1
            \   abcd
            |   e
            |___^

        "#}
```

**Review by orizi — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: all files reviewed, 5 unresolved discussions (waiting on @0xLucqs and @ilyalesokhin-starkware)

**Review by mkaput — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: all files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

**Review by mkaput — APPROVED** [review]
> Reviewed 14 of 14 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/6105)* status: all files reviewed, 5 unresolved discussions (waiting on @0xLucqs, @ilyalesokhin-starkware, and @orizi)

**Pull request closed without merging**